README.md
[![Build badge](https://img.shields.io/github/workflow/status/maikroempagel/quantonex/ElixirCI?label=Build)](https://github.com/maikroempagel/quantonex/actions?query=workflow%3AElixirCI)
[![Coveralls badge](https://img.shields.io/coveralls/github/maikroempagel/quantonex)](https://coveralls.io/github/maikroempagel/quantonex)
[![License badge](https://img.shields.io/github/license/maikroempagel/quantonex)](https://github.com/maikroempagel/quantonex/blob/master/LICENSE.txt)
[![Package badge](https://img.shields.io/hexpm/v/quantonex)](https://hex.pm/packages/quantonex)
# Quantonex

A technical analysis library for algorithmic trading.

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `quantonex` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:quantonex, "~> 0.1.2"}
  ]
end
```

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/quantonex](https://hexdocs.pm/quantonex).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](Https://conventionalcommits.org) for commit guidelines.

<!-- changelog -->

## [v0.1.2](https://github.com/maikroempagel/quantonex/compare/0.1.1...v0.1.2) (2020-03-15)




### Bug Fixes:

* rsi: Replace deprecated map operations

## [v0.1.1](https://github.com/maikroempagel/quantonex/compare/0.1.0...v0.1.1) (2020-03-14)




## [0.1.0](https://github.com/maikroempagel/quantonex/compare/0.1.0...0.1.0) (2020-02-29)


test/quantonex_test.exs
defmodule QuantonexTest do
  use ExUnit.Case

  @dataset_error "There must be at least 2 data points."
  @period_error "Period must be at least 2."

  test "invalid empty data set and implicit period" do
    assert Quantonex.sma([]) == {:error, @dataset_error}
  end

  test "invalid empty data set and explicit invalid period" do
    assert Quantonex.sma([], 1) == {:error, @dataset_error}
  end

  test "invalid empty data set and explicit valid period" do
    assert Quantonex.sma([], 2) == {:error, @dataset_error}
  end

  test "invalid non-empty data set and implicit period" do
    assert Quantonex.sma([1]) == {:error, @dataset_error}
  end

  test "valid non-empty data set and explicit invalid period" do
    assert Quantonex.sma([1, 2, 3], 1) == {:error, @period_error}
  end

  test "valid non-empty integer data set and implicit period" do
    assert Quantonex.sma([1, 2, 3]) == {:ok, Decimal.new(2)}
  end

  test "valid non-empty float data set and implicit period" do
    assert Quantonex.sma([2.1, 2.0, -2.0]) == {:ok, Decimal.from_float(0.7)}
  end

  test "invalid non-numeric data set" do
    data = ["a", "b"]
    assert Quantonex.sma(data) == {:error, "One of the data points is non-numeric."}
  end
end


test/indicators_test.exs
defmodule Quantonex.IndicatorsTest do
  @moduledoc false

  use ExUnit.Case, async: true

  alias Quantonex.DataPoint
  alias Quantonex.Indicators

  doctest Quantonex.Indicators

  @test_data_path Path.expand("../test/data", __DIR__)

  @dataset_min_size_error "There must be at least 1 element in the dataset."
  @period_min_value_error "Period must be at least 1."

  describe "ema/2" do
    test "empty dataset" do
      assert Indicators.ema([], 0) == {:error, @dataset_min_size_error}
    end

    test "period 0" do
      dataset = [22.81]

      assert Indicators.ema(dataset, 0) == {:error, @period_min_value_error}
    end

    test "period less than 0" do
      dataset = [22.81]

      assert Indicators.ema(dataset, -3) == {:error, @period_min_value_error}
    end

    test "non-numeric dataset" do
      invalid_element = "a"
      dataset = [invalid_element]

      error_message =
        "An error occured while calculating the EMA value: invalid_operation: number parsing syntax: #{
          invalid_element
        }"

      assert Indicators.ema(dataset, 1) == {:error, error_message}
    end

    test "verify dataset" do
      period = 10

      test_data = parse_test_data(:ema)

      dataset =
        test_data
        |> Enum.map(fn x ->
          {price, _ema} = x

          price
        end)

      {:ok, actual} = Indicators.ema(dataset, period)

      test_data
      |> Enum.with_index()
      |> Enum.each(fn x ->
        {{_price, expected_ema}, index} = x

        actual_ema = Enum.at(actual, index)

        assert Decimal.equal?(actual_ema, expected_ema),
               "Expected #{expected_ema}, but was #{actual_ema}!"
      end)
    end

    test "float dataset with period equal to the length of the dataset" do
      dataset = [
        22.2734,
        22.194,
        22.0847,
        22.1741,
        22.184,
        22.1344,
        22.2337,
        22.4323,
        22.2436,
        22.2933
      ]

      # this is equal to SMA(10)
      expected_results = [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        Decimal.new("22.22475")
      ]

      {:ok, actual} = Indicators.ema(dataset, 10)

      verify_results(actual, expected_results)
    end

    test "integer dataset with period less than length of dataset" do
      dataset = 1..10 |> Enum.map(fn x -> x end)

      expected_results = [
        0,
        0,
        0,
        0,
        Decimal.new(3),
        Decimal.new(4),
        Decimal.new(5),
        Decimal.new(6),
        Decimal.new(7),
        Decimal.new(8)
      ]

      {:ok, actual} = Indicators.ema(dataset, 5)

      verify_results(actual, expected_results)
    end

    test "string dataset with period less than length of dataset" do
      dataset = 1..10 |> Enum.map(fn x -> Integer.to_string(x) end)

      expected_results = [
        0,
        0,
        0,
        0,
        Decimal.new(3),
        Decimal.new(4),
        Decimal.new(5),
        Decimal.new(6),
        Decimal.new(7),
        Decimal.new(8)
      ]

      {:ok, actual} = Indicators.ema(dataset, 5)

      verify_results(actual, expected_results)
    end
  end

  describe "ema/3" do
    test "non-numeric price" do
      previous_ema = Decimal.from_float(5.5)
      invalid_price = "a"

      error_message =
        "An error occured while calculating the EMA value: invalid_operation: number parsing syntax: #{
          invalid_price
        }"

      assert Indicators.ema(invalid_price, 10, previous_ema) == {:error, error_message}
    end

    test "period 10" do
      previous_ema = Decimal.from_float(5.5)

      expected = Decimal.from_float(6.5)
      {:ok, actual} = Quantonex.Indicators.ema(11, 10, previous_ema)

      assert Decimal.equal?(actual, expected),
             "Expected #{expected}, but was #{actual}!"
    end
  end

  describe "rsi/2" do
    test "empty dataset" do
      dataset = []

      assert Indicators.rsi(dataset, 3) == {:error, @dataset_min_size_error}
    end

    test "non-numeric dataset" do
      invalid_element = "a"
      dataset = [invalid_element]

      error_message =
        "An error occured while calculating the RSI value: invalid_operation: number parsing syntax: #{
          invalid_element
        }"

      assert Indicators.rsi(dataset, 1) == {:error, error_message}
    end

    test "verify dataset" do
      period = 14

      test_data = parse_test_data(:rsi)

      dataset =
        test_data
        |> Enum.map(fn x ->
          {price, _rsi} = x

          price
        end)

      {:ok, actual} = Indicators.rsi(dataset, period)

      test_data
      |> Enum.with_index()
      |> Enum.each(fn x ->
        {{_price, expected_rsi}, index} = x

        actual_rsi = Enum.at(actual, index)

        assert Decimal.equal?(actual_rsi, expected_rsi),
               "Expected #{expected_rsi}, but was #{actual_rsi}!"
      end)
    end
  end

  describe "sma/1" do
    test "empty dataset" do
      assert Indicators.sma([]) == {:error, @dataset_min_size_error}
    end

    test "single element non-numeric dataset" do
      invalid_element = "a"
      dataset = [invalid_element]

      error_message =
        "An error occured while calculating the SMA value: invalid_operation: number parsing syntax: #{
          invalid_element
        }"

      assert Indicators.sma(dataset) == {:error, error_message}
    end

    test "float dataset" do
      dataset = [
        22.81,
        23.09,
        22.91,
        23.23,
        22.83,
        23.05,
        23.02,
        23.29,
        23.41
      ]

      expected = Decimal.new("23.07111111111111111111111111")

      {:ok, actual} = Indicators.sma(dataset)

      assert Decimal.equal?(actual, expected),
             "Expected #{expected}, but was #{actual}!"
    end

    test "integer dataset" do
      dataset =
        [
          101,
          100,
          103,
          99,
          96,
          99,
          95,
          91,
          93,
          89
        ]
        |> Enum.map(fn x -> x end)

      expected = Decimal.from_float(96.6)

      {:ok, actual} = Indicators.sma(dataset)

      assert Decimal.equal?(actual, expected),
             "Expected #{expected}, but was #{actual}!"
    end

    test "string dataset" do
      dataset =
        [
          101,
          100,
          103,
          99,
          96,
          99,
          95,
          91,
          93,
          89
        ]
        |> Enum.map(fn x -> Integer.to_string(x) end)

      expected = Decimal.from_float(96.6)

      {:ok, actual} = Indicators.sma(dataset)

      assert Decimal.equal?(actual, expected),
             "Expected #{expected}, but was #{actual}!"
    end
  end

  describe "vwap/1" do
    test "empty dataset" do
      dataset = []

      assert Indicators.vwap(dataset) == {:error, @dataset_min_size_error}
    end

    test "zero volume and cumulative volume" do
      data_point = %Quantonex.DataPoint{
        high: Decimal.from_float(127.36),
        low: Decimal.from_float(126.99),
        close: Decimal.from_float(127.28),
        volume: 0
      }

      error_message = "The data point volume and cumulative volume can't both be zero."

      assert Indicators.vwap([data_point]) == {:error, error_message}
    end

    test "invalid data point" do
      invalid_element = "a"

      data_point = %Quantonex.DataPoint{
        low: invalid_element,
        volume: 10
      }

      error_message =
        "An error occured while calculating the VWAP value: invalid_operation: number parsing syntax: #{
          invalid_element
        }"

      assert Quantonex.Indicators.vwap(data_point) ==
               {:error, error_message}
    end

    test "verify dataset" do
      test_data = parse_test_data(:vwap)

      dataset =
        test_data
        |> Enum.map(fn x ->
          {data_point, _vwap} = x

          data_point
        end)

      {:ok, actual} = Indicators.vwap(dataset)

      test_data
      |> Enum.with_index()
      |> Enum.each(fn x ->
        {{_data_point, expected_vwap}, index} = x

        actual_vwap = Enum.at(actual, index)

        assert actual_vwap.cumulative_volume == expected_vwap.cumulative_volume

        assert Decimal.equal?(
                 actual_vwap.cumulative_volume_price,
                 expected_vwap.cumulative_volume_price
               )

        assert Decimal.equal?(
                 actual_vwap.value,
                 expected_vwap.value
               )
      end)
    end
  end

  describe "vwap/3" do
    test "zero volume and cumulative volume" do
      data_point = %Quantonex.DataPoint{
        volume: 0
      }

      assert Quantonex.Indicators.vwap(data_point, 0, 1) ==
               {:error, "The data point volume and cumulative volume can't both be zero."}
    end

    test "negative cumulative volume" do
      data_point = %Quantonex.DataPoint{}

      assert Quantonex.Indicators.vwap(data_point, -1, 1) ==
               {:error, "The cumulative volume can't be negative."}
    end

    test "negative cumulative volume price" do
      data_point = %Quantonex.DataPoint{
        volume: 10
      }

      assert Quantonex.Indicators.vwap(data_point, 0, -1) ==
               {:error, "The cumulative volume price can't be negative."}
    end

    test "invalid data point" do
      invalid_element = "a"

      data_point = %Quantonex.DataPoint{
        low: invalid_element
      }

      error_message =
        "An error occured while calculating the VWAP value: invalid_operation: number parsing syntax: #{
          invalid_element
        }"

      assert Quantonex.Indicators.vwap(data_point, 2, 2) ==
               {:error, error_message}
    end

    test "data point with implicit cumulative volume price" do
      data_point = %Quantonex.DataPoint{
        complete: false,
        high: Decimal.from_float(127.36),
        low: Decimal.from_float(126.99),
        close: Decimal.from_float(127.28),
        volume: 89_329
      }

      expected_vwap = %{
        cumulative_volume: 89_329,
        cumulative_volume_price: Decimal.from_float(11_363_542.09),
        value: Decimal.from_float(127.21)
      }

      assert Quantonex.Indicators.vwap(data_point, 0) == {:ok, expected_vwap}
    end

    test "data point with implicit cumulative values" do
      data_point = %Quantonex.DataPoint{
        complete: false,
        high: Decimal.new(8),
        low: Decimal.new(4),
        close: Decimal.new(6),
        volume: 10
      }

      expected_vwap = %{
        cumulative_volume: 10,
        cumulative_volume_price: Decimal.new(60),
        value: Decimal.new(6)
      }

      assert Quantonex.Indicators.vwap(data_point) == {:ok, expected_vwap}
    end

    test "data point and cumulative values" do
      data_point = %Quantonex.DataPoint{
        complete: false,
        high: Decimal.new(12),
        low: Decimal.new(8),
        close: Decimal.new(10),
        volume: 20
      }

      expected_vwap = %{
        cumulative_volume: 30,
        cumulative_volume_price: Decimal.new(240),
        value: Decimal.new(8)
      }

      assert Quantonex.Indicators.vwap(data_point, 10, Decimal.new(40)) == {:ok, expected_vwap}
    end
  end

  ## Helpers

  defp lines(filename) do
    Path.join([@test_data_path, filename])
    |> File.read!()
    |> String.split("\n", trim: true)
  end

  # used as a guard against parse errors
  defp map_lines([], _fun), do: nil

  defp map_lines(lines, fun) do
    lines
    |> Enum.slice(1..(length(lines) - 1))
    |> Enum.map(&String.split(&1, ","))
    |> Enum.map(fn x -> fun.(x) end)
  end

  defp parse_test_data(:ema) do
    # test data taken from: https://school.stockcharts.com/doku.php?id=technical_indicators:vwap_intraday

    lines("ema_test_data")
    |> map_lines(fn x ->
      price = Decimal.new(Enum.at(x, 0))
      ema = Decimal.new(Enum.at(x, 1))

      {price, ema}
    end)
  end

  defp parse_test_data(:rsi) do
    # test data taken from: https://school.stockcharts.com/doku.php?id=technical_indicators:moving_averages

    lines("rsi_test_data")
    |> map_lines(fn x ->
      price = Decimal.new(Enum.at(x, 0))
      rsi = Decimal.new(Enum.at(x, 1))

      {price, rsi}
    end)
  end

  defp parse_test_data(:vwap) do
    # test data taken from: https://school.stockcharts.com/doku.php?id=technical_indicators:vwap_intraday

    lines("vwap_test_data")
    |> map_lines(fn x ->
      data_point = %DataPoint{
        high: Decimal.new(Enum.at(x, 0)),
        low: Decimal.new(Enum.at(x, 1)),
        close: Decimal.new(Enum.at(x, 2)),
        volume: String.to_integer(Enum.at(x, 3))
      }

      vwap = %{
        cumulative_volume: String.to_integer(Enum.at(x, 4)),
        cumulative_volume_price: Decimal.new(Enum.at(x, 5)),
        value: Decimal.new(Enum.at(x, 6))
      }

      {data_point, vwap}
    end)
  end

  defp verify_results(actual_results, expected_results) do
    actual_results
    |> Enum.with_index()
    |> Enum.each(fn x ->
      {value, index} = x

      expected = Enum.at(expected_results, index)

      assert Decimal.equal?(value, expected),
             "Expected #{expected}, but was #{value}!"
    end)
  end
end


mix.exs
defmodule Quantonex.MixProject do
  use Mix.Project

  @github_url "https://github.com/maikroempagel/quantonex"

  @version "0.1.2"

  def project do
    [
      app: :quantonex,
      version: @version,
      elixir: "~> 1.9",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      docs: docs(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      source_url: @github_url,
      homepage_url: @github_url,
      files: ~w(mix.exs lib LICENSE.txt README.md CHANGELOG.md),
      package: package()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.2", only: [:dev, :test], runtime: false},
      {:decimal, "~> 1.8"},
      {:ex_doc, "~> 0.21.3", only: :dev, runtime: false},
      {:excoveralls, "~> 0.12.2", only: :test},
      {:git_ops, "~> 1.1.1", only: :dev}
    ]
  end

  defp description() do
    """
    A technical analysis library for algorithmic trading.
    """
  end

  defp docs, do: []

  defp package() do
    [
      name: :quantonex,
      maintainers: ["Maik Römpagel"],
      licenses: ["Apache-2.0"],
      links: %{
        "GitHub" => @github_url
      }
    ]
  end
end


lib/quantonex.ex
defmodule Quantonex do
  @moduledoc """
  Documentation for Quantonex.
  """

  def sma(data) do
    data |> sma(length(data))
  end

  def sma(data, _period) when length(data) < 2,
    do: {:error, "There must be at least 2 data points."}

  def sma(_data, period) when period < 2, do: {:error, "Period must be at least 2."}

  def sma(data, period) do
    try do
      value =
        data
        |> Enum.reverse()
        |> Enum.take(period)
        |> Enum.map(&create_decimal/1)
        |> Enum.reduce(fn x, acc -> Decimal.add(x, acc) end)
        |> Decimal.div(period)

      {:ok, value}
    rescue
      _e in Decimal.Error ->
        {:error, "One of the data points is non-numeric."}
    end
  end

  defp create_decimal(value) when is_float(value), do: Decimal.from_float(value)

  defp create_decimal(value), do: Decimal.new(value)
end


lib/indicators.ex
defmodule Quantonex.Indicators do
  @moduledoc """
  Contains technical indicators.
  """

  alias Quantonex.DataPoint

  @dataset_min_size_error "There must be at least 1 element in the dataset."
  @period_min_value_error "Period must be at least 1."
  @period_max_value_error "Period can't be greater than the length of the dataset."

  @zero Decimal.new(0)

  @typedoc """
  Represents a smoothing method.

    * `:ema` - exponential moving average
    * `:sma` - simple moving average
  """
  @type smoothing_method :: :ema | :sma

  @doc """
  Calculates a list of exponential moving averages (EMAs) for a given dataset and period.

  The first `n` elements (`n == period`) are used to calculate the initial EMA using a SMA.
  Each successive value is calculated using an EMA.

  Possible return values are:

  * `{:error, reason}`
  * `{:ok, values}`

  The returned list of EMA values has the same length as the input dataset, so they can
  be joined again.

  ## Examples

    ```
    dataset = 1..11 |> Enum.map(fn x -> x end)
    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

    {:ok, emas} = Quantonex.Indicators.ema(dataset, 10)
    {:ok,
    [#Decimal<0>, #Decimal<0>, #Decimal<0>, #Decimal<0>, #Decimal<0>, #Decimal<0>,
    #Decimal<0>, #Decimal<0>, #Decimal<0>, #Decimal<5.5>,
    #Decimal<6.500000000000000000000000000>]}

    Enum.zip(dataset, emas)
    [
      {1, #Decimal<0>},
      {2, #Decimal<0>},
      {3, #Decimal<0>},
      {4, #Decimal<0>},
      {5, #Decimal<0>},
      {6, #Decimal<0>},
      {7, #Decimal<0>},
      {8, #Decimal<0>},
      {9, #Decimal<0>},
      {10, #Decimal<5.5>},
      {11, #Decimal<6.500000000000000000000000000>}
    ]
    ```
  """
  @spec ema(dataset :: nonempty_list(price :: String.t() | number()), period :: pos_integer()) ::
          {:error, reason :: String.t()} | {:ok, nonempty_list(values :: Decimal.t())}

  def ema([], _period), do: {:error, @dataset_min_size_error}

  def ema(dataset, period) when is_list(dataset) and period < 1,
    do: {:error, @period_min_value_error}

  def ema(dataset, period) when is_list(dataset) and period > length(dataset),
    do: {:error, @period_max_value_error}

  def ema(dataset, period) when is_list(dataset) do
    try do
      # every ema value with an index < period is set to 0
      initial_emas = 1..(period - 1) |> Enum.map(fn _ -> @zero end)

      # the first EMA is based on a SMA
      {:ok, seed} =
        dataset
        |> Enum.take(period)
        |> Enum.map(&to_decimal/1)
        |> sma()

      values =
        dataset
        |> Enum.slice(period..(length(dataset) - 1))
        |> Enum.map(&to_decimal/1)
        |> Enum.reduce_while([seed | initial_emas], fn current_price, acc ->
          [previous_ema | _tail] = acc

          case ema(current_price, period, previous_ema) do
            {:ok, value} -> {:cont, [value | acc]}
            {:error, reason} -> {:halt, {:error, reason}}
          end
        end)
        # reverse to match the order of the input dataset
        |> Enum.reverse()

      {:ok, values}
    rescue
      e in Decimal.Error ->
        {:error, "An error occured while calculating the EMA value: " <> e.message}
    end
  end

  @doc """
  Calculates a single exponential moving average (EMA) for a given price, period and another EMA.

  This is a convenience function to avoid calculation using a complete dataset.
  See `Quantonex.Indicators.ema/2` for how to calculate an initial EMA.

  Possible return values are:

  * `{:error, reason}`
  * `{:ok, value}`

  ## Examples

    ```
    previous_ema = Decimal.from_float(5.5)
    #Decimal<5.5>

    {:ok, value} = Quantonex.Indicators.ema(11, 10, previous_ema)
    {:ok, #Decimal<6.500000000000000000000000000>}
    ```
  """
  @spec ema(price :: String.t() | number(), period :: pos_integer(), previous_ema :: Decimal.t()) ::
          {:error, reason :: String.t()} | {:ok, value :: Decimal.t()}
  def ema(price, period, previous_ema) do
    try do
      # 2 / (period + 1)
      multiplier = weighted_multiplier(period)

      # (price - previous_ema) * multiplier + previous_ema
      value =
        previous_ema
        |> to_decimal()
        |> Decimal.mult(Decimal.sub(Decimal.new(1), multiplier))
        |> Decimal.add(Decimal.mult(price, multiplier))

      {:ok, value}
    rescue
      e in Decimal.Error ->
        {:error, "An error occured while calculating the EMA value: " <> e.message}
    end
  end

  @doc """
  Calculates a list of relative strength indexes (RSIs) for a given dataset.
  """
  @spec rsi(
          dataset :: nonempty_list(price :: String.t() | number()),
          period :: non_neg_integer()
        ) ::
          {:error, reason :: String.t()}
          | {:ok, values :: nonempty_list(Decimal.t())}

  def rsi([], _period), do: {:error, @dataset_min_size_error}

  def rsi(dataset, period) do
    try do
      init_map = %{
        current_price: @zero,
        previous_price: @zero,
        up_movement: @zero,
        up_sum: @zero,
        down_movement: @zero,
        down_sum: @zero,
        up_average: @zero,
        down_average: @zero,
        relative_strength: @zero,
        value: @zero
      }

      results =
        dataset
        |> Enum.map(&to_decimal/1)
        |> Enum.with_index()
        |> Enum.reduce(fn price_with_index, acc ->
          {current_price, index} = price_with_index

          {previous_rsi, previous_items} =
            case acc do
              # the first iteration and initialization of our result rsi dataset
              {first_price, _index} ->
                previous_rsi = %{init_map | current_price: first_price}

                {
                  previous_rsi,
                  [previous_rsi]
                }

              [previous_rsi | _tail] ->
                {previous_rsi, acc}
            end

          current_rsi =
            init_map
            |> Map.put(:current_price, current_price)
            |> Map.put(:previous_price, previous_rsi.current_price)
            |> rsi_up_movement()
            |> rsi_down_movement()
            |> rsi_up_sum(previous_rsi)
            |> rsi_down_sum(previous_rsi)
            |> rsi_up_average(previous_rsi, index, period)
            |> rsi_down_average(previous_rsi, index, period)
            |> rsi_strength(index, period)
            |> rsi_index(index, period)

          [current_rsi | previous_items]
        end)
        |> Enum.map(fn x -> x.value end)
        |> Enum.reverse()

      {:ok, results}
    rescue
      e in Decimal.Error ->
        {:error, "An error occured while calculating the RSI value: " <> e.message}
    end
  end

  @doc """
  Calculates a simple moving average (SMA) for a given dataset.

  The period of the SMA is fixed and equal to the length of the dataset.

  Possible return values are:

  * `{:error, reason}`
  * `{:ok, value}`

  ## Examples

    ```
    iex> Quantonex.Indicators.sma([1, 2, 3])
    {:ok, Decimal.new(2)}
    ```
  """
  @spec sma(dataset :: nonempty_list(price :: String.t() | number())) ::
          {:error, reason :: String.t()} | {:ok, value :: Decimal.t()}
  def sma([]), do: {:error, @dataset_min_size_error}

  def sma(dataset) when is_list(dataset) do
    try do
      value = calculate_simple_moving_average(dataset, length(dataset))

      {:ok, value}
    rescue
      e in Decimal.Error ->
        {:error, "An error occured while calculating the SMA value: " <> e.message}
    end
  end

  @typedoc """
  Represents a volume weighted average price.

    * `value` - the volume weighted average price
    * `cumulative_volume` - the previous volume plus the current volume
    * `cumulative_volume_price` - the previous volume price plus the current volume price
  """
  @type volume_weighted_average_price :: %{
          cumulative_volume: non_neg_integer(),
          cumulative_volume_price: Decimal.t(),
          value: Decimal.t()
        }

  @doc """
  Calculates a list of volume weighted average prices (VWAPs) for a given dataset.

  The following data point properties are used to calculate a VWAP value.

  * `high`
  * `low`
  * `close`
  * `volume`

  Possible return values are:

  * `{:error, reason}`
  * `{:ok, values}`

  The returned list of VWAP values has the same length as the input dataset, so they can
  be joined again.

  ## Examples

    ```
    dataset = [%Quantonex.DataPoint{
                  close: #Decimal<127.28>,
                  high: #Decimal<127.36>,
                  low: #Decimal<126.99>,
                  volume: 89329
                },...]

    {:ok, vwaps} = Quantonex.Indicators.vwap(dataset)
    {:ok,
      [%{
        cumulative_volume: 89329,
        cumulative_volume_price: #Decimal<11363542.09>,
        value: #Decimal<127.21>
      }, ...]
    }

    Enum.zip(dataset, vwaps)
    [
      ...
    ]
    ```
  """
  @spec vwap(dataset :: nonempty_list(DataPoint.t())) ::
          {:error, reason :: String.t()}
          | {:ok, values :: nonempty_list(volume_weighted_average_price())}
  def vwap([]), do: {:error, @dataset_min_size_error}

  def vwap(dataset) when is_list(dataset) do
    values =
      dataset
      |> Enum.reduce_while([], fn data_point, acc ->
        {cumulative_volume, cumulative_volume_price} =
          case acc do
            [] ->
              {0, @zero}

            [previous_vwap | _tail] ->
              {previous_vwap.cumulative_volume, previous_vwap.cumulative_volume_price}
          end

        case vwap(data_point, cumulative_volume, cumulative_volume_price) do
          {:ok, value} -> {:cont, [value | acc]}
          {:error, reason} -> {:halt, {:error, reason}}
        end
      end)

    case values do
      {:error, reason} -> {:error, reason}
      _ -> {:ok, values |> Enum.reverse()}
    end
  end

  @doc """
  Calculates a single volume weighted average price (VWAP).

  The following data point properties are used to calculate a VWAP value.

  * `high`
  * `low`
  * `close`
  * `volume`

  Possible return values are:

  * `{:error, reason}`
  * `{:ok, values}`

  ## Examples

  ```
  data_point = %Quantonex.DataPoint{
    close: Decimal.new(6),
    high: Decimal.new(8),
    low: Decimal.new(4),
    volume: 10
  }

  {:ok,
    %{
      "cumulative_volume": cumulative_volume,
      "cumulative_volume_price": cumulative_volume_price,
      "value": value
    }
  } = Quantonex.Indicators.vwap(data_point)
  ```

  Successive calculations can be done by passing previously calculated cumulative values to
  the function.

  ```
  next_data_point = %Quantonex.DataPoint{
    close: Decimal.new(8),
    high: Decimal.new(10),
    low: Decimal.new(6),
    volume: 20
  }

  Quantonex.Indicators.vwap(next_data_point, cumulative_volume, cumulative_volume_price)
  ```
  """
  @spec vwap(
          data_point :: DataPoint.t(),
          cumulative_volume :: non_neg_integer(),
          cumulative_volume_price :: number() | Decimal.t()
        ) ::
          {:error, reason :: String.t()} | {:ok, value :: volume_weighted_average_price()}

  def vwap(data_point, cumulative_volume \\ 0, cumulative_volume_price \\ 0)

  def vwap(%DataPoint{:volume => volume}, cumulative_volume, _cumulative_volume_price)
      when volume == 0 and cumulative_volume == 0,
      do: {:error, "The data point volume and cumulative volume can't both be zero."}

  def vwap(%DataPoint{}, cumulative_volume, _cumulative_volume_price)
      when is_integer(cumulative_volume) and cumulative_volume < 0,
      do: {:error, "The cumulative volume can't be negative."}

  def vwap(%DataPoint{}, _cumulative_volume, cumulative_volume_price)
      when cumulative_volume_price < 0,
      do: {:error, "The cumulative volume price can't be negative."}

  def vwap(%DataPoint{} = data_point, cumulative_volume, cumulative_volume_price) do
    try do
      # ensure cumulative volume price can be converted to a decimal
      cumulative_volume_price = to_decimal(cumulative_volume_price)

      average_price =
        data_point.high
        |> Decimal.add(data_point.low)
        |> Decimal.add(data_point.close)
        |> Decimal.div(3)

      volume_price = Decimal.mult(average_price, data_point.volume)

      new_cumulative_volume = cumulative_volume + data_point.volume
      new_cumulative_volume_price = Decimal.add(cumulative_volume_price, volume_price)

      value = Decimal.div(new_cumulative_volume_price, new_cumulative_volume)

      {
        :ok,
        %{
          cumulative_volume: new_cumulative_volume,
          cumulative_volume_price: new_cumulative_volume_price,
          value: value
        }
      }
    rescue
      e in Decimal.Error ->
        {:error, "An error occured while calculating the VWAP value: " <> e.message}
    end
  end

  ## Helpers

  defp calculate_simple_moving_average(dataset, period) do
    dataset
    |> Enum.map(&to_decimal/1)
    |> Enum.reduce(fn x, acc -> Decimal.add(x, acc) end)
    |> calculate_average(period)
  end

  defp calculate_average(sum, divisor), do: Decimal.div(sum, Decimal.new(divisor))

  defp rsi_down_average(current_rsi, _previous_rsi, index, period) when index < period,
    do: current_rsi

  defp rsi_down_average(%{:down_sum => down_sum} = current_rsi, _previous_rsi, index, period)
       when index == period do
    %{current_rsi | down_average: calculate_simple_moving_average([down_sum], period)}
  end

  defp rsi_down_average(
         %{:down_movement => current_down_movement} = current_rsi,
         %{
           :down_average => previous_down_average
         } = _previous_rsi,
         _index,
         period
       ) do
    down_average =
      Decimal.mult(previous_down_average, Decimal.new(period - 1))
      |> Decimal.add(current_down_movement)
      |> Decimal.div(Decimal.new(period))

    %{current_rsi | down_average: down_average}
  end

  defp rsi_index(rsi_map, index, period) when index < period, do: rsi_map

  defp rsi_index(%{:relative_strength => relative_strength} = rsi_map, _index, _period) do
    max_rsi = Decimal.new(100)

    # RSI: 100 – 100 / ( 1 + relative_strength)
    first_calc = Decimal.new(1) |> Decimal.add(relative_strength)
    second_calc = Decimal.div(max_rsi, first_calc)

    relative_strength_index = Decimal.sub(max_rsi, second_calc)

    %{rsi_map | value: relative_strength_index}
  end

  defp rsi_strength(%{:down_average => @zero} = rsi_map, index, period) when index < period,
    do: rsi_map

  defp rsi_strength(
         %{
           :up_average => up_average,
           :down_average => down_average
         } = rsi_map,
         _index,
         _period
       ) do
    %{rsi_map | relative_strength: Decimal.div(up_average, down_average)}
  end

  defp rsi_up_sum(
         %{:up_movement => up_movement} = current_rsi,
         %{:up_sum => up_sum} = _previous_rsi
       ) do
    up_sum = Decimal.add(up_movement, up_sum)

    %{current_rsi | up_sum: up_sum}
  end

  defp rsi_down_sum(
         %{:down_movement => down_movement} = current_rsi,
         %{:down_sum => down_sum} = _previous_rsi
       ) do
    down_sum = Decimal.add(down_movement, down_sum)

    %{current_rsi | down_sum: down_sum}
  end

  defp rsi_down_movement(
         %{:previous_price => previous_price, :current_price => current_price} = rsi_map
       ) do
    diff = Decimal.sub(current_price, previous_price)

    case Decimal.lt?(diff, 0) do
      true -> %{rsi_map | down_movement: Decimal.abs(diff)}
      false -> %{rsi_map | down_movement: @zero}
    end
  end

  defp rsi_up_average(current_rsi, _previous_rsi, index, period) when index < period,
    do: current_rsi

  defp rsi_up_average(%{:up_sum => up_sum} = current_rsi, _previous_rsi, index, period)
       when index == period do
    %{current_rsi | up_average: calculate_simple_moving_average([up_sum], period)}
  end

  defp rsi_up_average(
         %{:up_movement => current_up_movement} = current_rsi,
         %{
           :up_average => previous_up_average
         } = _previous_rsi,
         _index,
         period
       ) do
    up_average =
      Decimal.mult(previous_up_average, Decimal.new(period - 1))
      |> Decimal.add(current_up_movement)
      |> Decimal.div(Decimal.new(period))

    %{current_rsi | up_average: up_average}
  end

  defp rsi_up_movement(
         %{:previous_price => previous_price, :current_price => current_price} = rsi_map
       ) do
    diff = Decimal.sub(current_price, previous_price)

    case Decimal.gt?(diff, 0) do
      true -> %{rsi_map | up_movement: diff}
      false -> %{rsi_map | up_movement: @zero}
    end
  end

  defp to_decimal(value) when is_float(value), do: Decimal.from_float(value)
  # create decimals from either strings, integers or decimals
  defp to_decimal(value), do: Decimal.new(value)

  defp weighted_multiplier(period) do
    period_increment = Decimal.new(period + 1)

    Decimal.new(2)
    |> Decimal.div(period_increment)
  end
end
